strawberryfields/api_client.py

```diff
@@ -0,0 +1,546 @@
+
+        if response.status_code in (400, 409):
+            warnings.warn(
+                "The server did not accept the request, and returned an error "
+                "({}: {}).".format(response.status_code, response.text),
+                UserWarning,
+            )
+        elif response.status_code == 401:
+            warnings.warn(
+                "The server did not accept the request due to an authentication error "
+                "({}: {}).".format(response.status_code, response.text),
+                UserWarning,
+            )
+        elif response.status_code in (500, 503, 504):
+            warnings.warn(
+                "The client encountered an unexpected temporary server error "
+                "({}: {}).".format(response.status_code, response.text),
+                UserWarning,
+            )
+
+    def refresh_data(self, data):
+        """
+        Refreshes the instance's attributes with the provided data and
+        converts it to the correct type.
+
+        Args:
+            data (dict): A dictionary containing keys and values of data to be stored on the object.
+        """
+        for field in self.resource.fields:
+            field.set(data.get(field.name, None))
+
+
+class Resource:
+    """
+    A base class for an API resource. This class should be extended for each
+    resource endpoint.
+    """
+
+    SUPPORTED_METHODS = ()
+    PATH = ""
+    fields = ()
+
+    def __init__(self, client=None):
+        """
+        Initialize the Resource by populating attributes based on fields and settings a manager.
+
+        Args:
+            client (APIClient): An APIClient instance to use as a client.
+        """
+        self.manager = ResourceManager(self, client=client)
+        for field in self.fields:
+            setattr(self, field.name, field)
+
+    def reload(self):
+        """
+        A helper method to fetch the latest data from the API.
+        """
+        if not hasattr(self, "id"):
+            raise TypeError("Resource does not have an ID")
+
+        if self.id:
+            self.manager.get(self.id)
+        else:
+            warnings.warn("Could not reload resource data", UserWarning)
+
+
+class Field:
+    """
+    A helper class to classify and clean data returned by the API.
+    """
+
+    value = None
+
+    def __init__(self, name, clean=str):
+        """
+        Initialize the Field object with a name and a cleaning function.
+
+        Args:
+            name (str): A string representing the name of the field (e.g. "created_at").
+            clean: A method that returns a cleaned value of the field, of the correct type.
+        """
+        self.name = name
+        self.clean = clean
+
+    def __str__(self):
+        """
+        Return the string representation of the value.
+        """
+        return str(self.value)
+
+    def __bool__(self):
+        """
+        Use the value to determine boolean state.
+        """
+        return self.value is not None
```

Comment:
I guess this is determining if there is a non-`None` state. 
So you could have a Field F whose value is `False`, but `__bool__` would still return `True`, correct?

Reply:
Yes that is correct. I probably should use a dirty flag here, or refactor the code to have a `Field.is_set` boolean value.